lib/Dialect/Substrait/IR/Substrait.cpp

```diff
@@ -1035,6 +1035,107 @@ LiteralOp::inferReturnTypes(MLIRContext *context, std::optional<Location> loc,
   return success();
 }
 
+LogicalResult
+HashJoinOp::inferReturnTypes(MLIRContext *context, std::optional<Location> loc,
+                             ValueRange operands, DictionaryAttr attributes,
+                             OpaqueProperties properties, RegionRange regions,
+                             llvm::SmallVectorImpl<Type> &inferredReturnTypes) {
+  Value leftInput = operands[0];
+  Value rightInput = operands[1];
+
+  TypeRange leftFieldTypes = cast<TupleType>(leftInput.getType()).getTypes();
+  TypeRange rightFieldTypes = cast<TupleType>(rightInput.getType()).getTypes();
+
+  // Get accessor to `join_type`.
+  Adaptor adaptor(operands, attributes, properties, regions);
+  JoinType join_type = adaptor.getJoinType();
+
+  SmallVector<mlir::Type> fieldTypes;
+
+  switch (join_type) {
+  case JoinType::unspecified:
+  case JoinType::inner:
+  case JoinType::outer:
+  case JoinType::right:
+  case JoinType::left:
+    llvm::append_range(fieldTypes, leftFieldTypes);
+    llvm::append_range(fieldTypes, rightFieldTypes);
+    break;
+  case JoinType::semi:
+  case JoinType::anti:
+    llvm::append_range(fieldTypes, leftFieldTypes);
+    break;
+  case JoinType::single:
+    llvm::append_range(fieldTypes, rightFieldTypes);
+    break;
+  }
+
+  auto resultType = TupleType::get(context, fieldTypes);
+
+  inferredReturnTypes = SmallVector<Type>{resultType};
+
+  return success();
+}
+
+LogicalResult KeyComparisonOp::verify() {
+  auto &res =
+      llvm::TypeSwitch<mlir::Type, LogicalResult>(getLhs().getType())
+          .Case<substrait::VarCharType, substrait::StringType>([&](auto) {
+            if (!mlir::isa<substrait::VarCharType, substrait::StringType>(
+                    getRhs()))
+              return this->emitError("Invalid rhs type for string comparison, "
+                                     "expected string-like type but got")
+                     << getRhs().getType();
+            return success();
+          })
+          .Case<IntegerType>([&](auto) {
+            if (!mlir::isa<IntegerType, substrait::DecimalType>(getRhs()))
+              return this->emitError("Invalid rhs type for integer comparison, "
+                                     "expected integer-like type but got")
+                     << getRhs().getType();
+            return success();
+          });
+  if (failed(res))
+    return res;
+
+  return success();
+}
+
+LogicalResult HashJoinOp::verifyRegions() {
+  if (getCondition().empty())
+    return emitOpError() << "hash join must have a condition region";
+
+  // Verify that we have exactly two arguments in the region, matching the left
+  // and right relations
+  Block &block = getCondition().front();
```

Comment:
Since you've added the `SingleBlockImplicitTerminator` trait, you should be able to get the body block by just saying `Block* body = getBody()`.